### PR TITLE
Model

### PR DESCRIPTION
**Pull request by svince**
Refactored Gideons model to match the stanford template.
There is still stuff to be done.

### CODE REVIEW

File: lstm.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+
+    if weight_decay is not None:
+        wd = tf.mul(tf.nn.l2_loss(var), weight_decay, name='weight_loss')
+        tf.add_to_collection('losses', wd)
+
+    return var
+
+
+  def load_data(self):
+    print "loading data.."
```
Comment (gideonite) [inline lstm.py:39]: Nitpick: please use python >=3.0. Ok so the pythonistas are pretty generous -- python2 only dies in 2020 -- but we might as well switch now. https://pythonclock.org/

File: main.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
+import tensorflow as tf
+from lstm import Config
+from lstm import Lstm
+
+def test_LstmModel():
+  config = Config()
+  config.data_path = "data/sentences_debug.train"
```
Comment (gideonite) [inline main.py:7]: Why is this hard coded? I thought we were using commandline parameters for this stuff. Also, this file is not standard...

File: lstm.py [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+    data_reader = DataReader()
+    data_reader.construct(self.config.data_path,
+      self.config.vocab_size, self.config.sentence_length)
+    return data_reader
+
+
+  def add_placeholders(self):
+
```
Comment (gideonite) [inline lstm.py:47]: nitpick. I know this is probably the default spacing from Stanford but can we all agree that it's stupid and it sucks? I vote - get rid of this extra whitespace.

File: lstm.py [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+
+  def add_embedding(self, input_data):
+    """Add embedding layer that maps from vocabulary to vectors.
+    Args:
+      input_data: A tensor of shape (batch_size, sentence_length).
+    Returns:
+      wordvectors: A tensor of shape (batch_size, sentence_length, embed_dim)
+    """
+    with tf.variable_scope('foobar'): # TODO this is stupid
```
Comment (gideonite) [inline lstm.py:64]: s/foobar/embedding/
Delete the TODO.

File: lstm.py [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+    """Implements core of the model.
+    Args:
+      input_data: A tensor of shape (batch_size, sentence_length).
+    Returns:
+      logitss: A tensor of shape (batch_size, sentence_length, vocab_size)
+	   For each batch, there are the logits of for each word of the sentence
+    """
+    lstm = tf.nn.rnn_cell.BasicLSTMCell(self.config.state_size)
+
+    with tf.variable_scope('foobar'): # TODO this is stupid
```
Comment (gideonite) [inline lstm.py:82]: s/foobar/softmax_layer/

File: lstm.py [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+        softmax_b = self._get_variable('softmax_b', [self.config.vocab_size])
+
+    wordvectors = self.add_embedding(input_data)
+
+    memory_state = tf.Variable(tf.zeros(
+      [self.config.batch_size, self.config.state_size]))
+    hidden_state = tf.Variable(tf.zeros(
+      [self.config.batch_size, self.config.state_size]))
+    state = (memory_state, hidden_state)
+    logitss = []
```
Comment (gideonite) [inline lstm.py:94]: better name. "sentence_logits"?

File: lstm.py [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+    '''
+    start_time = time.time()
+    predictions = []
+    
+    for i, batch in input_data.get_iterator(self.config.batch_size):
+      feed_dict = self.create_feed_dict(batch)
+      # TODO: Maybe we cannot use the same "model" as in the learning,
+      #       because there we always feed in the ground truth, and not the
+      #       output word of the previous step.
+      logitss, loss_value = sess.run([self.logitss, self.loss], feed_dict=feed_dict)
```
Comment (gideonite) [inline lstm.py:200]: Indeed, I think that for prediction we don't need a loss either.

**Review by gideonite — COMMENTED** [review]
LGTM as soon as I can get it to work but I haven't been able to because of a few minor bugs -- data directories and python 2 vs 3. We need to agree on a python version. I vote python 3 because...Guido van Rossum. But, if you insist on living in the stone age, that's fine, I'll suck it up.

I would have made the minor changes myself but I am hesitant to do so without being able to get it to work...

Also, is the code directory dead?
